From 59def3793080ac6b4f00457c87fa4fda0a328734 Mon Sep 17 00:00:00 2001
From: Wen Congyang <wency@cn.fujitsu.com>
Date: Mon, 16 Mar 2015 17:14:59 +0800
Subject: [PATCH 15/18] Don't allow a disk use backing reference target

Signed-off-by: Wen Congyang <wency@cn.fujitsu.com>
Signed-off-by: zhanghailiang <zhang.zhanghailiang@huawei.com>
Signed-off-by: Gonglei <arei.gonglei@huawei.com>
---
 block.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/block.c b/block.c
index 2664f13..d7c843d 100644
--- a/block.c
+++ b/block.c
@@ -1307,6 +1307,14 @@ static int bdrv_open_backing_reference_file(BlockDriverState *bs,
     }
 
     backing_hd = blk_bs(backing_blk);
+    /* Don't allow a disk use backing reference target */
+    ret = blk_attach_dev(backing_hd->blk, bs);
+    if (ret < 0) {
+        error_setg(errp, "backing_hd %s is used by the other device model",
+                   backing_name);
+        goto free_exit;
+    }
+
     /* Backing reference itself? */
     if (backing_hd == bs || bdrv_find_overlay(backing_hd, bs)) {
         error_setg(errp, "Backing reference itself");
@@ -1983,6 +1991,7 @@ void bdrv_close(BlockDriverState *bs)
                 assert(backing_hd->backing_hd);
                 if (backing_hd->backing_hd->job)
                     block_job_cancel(backing_hd->backing_hd->job);
+                blk_detach_dev(backing_hd->backing_hd->blk, bs);
                 bdrv_set_backing_hd(backing_hd, NULL);
                 bdrv_unref(backing_hd->backing_hd);
             }
-- 
1.7.1

